types.go

```diff
@@ -0,0 +1,296 @@
+package autopilot
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/raft"
+)
+
+//go:generate mockery -all -inpkg -case snake -testonly
+
+// RaftState is the status of a single server in the Raft cluster.
+type RaftState string
+
+const (
+	RaftNone     RaftState = "none"
+	RaftLeader   RaftState = "leader"
+	RaftVoter    RaftState = "voter"
+	RaftNonVoter RaftState = "non-voter"
+	RaftStaging  RaftState = "staging"
+)
+
+func (s RaftState) IsPotentialVoter() bool {
+	switch s {
+	case RaftVoter, RaftStaging, RaftLeader:
+		return true
+	default:
+		return false
+	}
+}
+
+// NodeStatus represents the health of a server as know to the autopilot consumer.
+// This should not take into account Raft health and the server being on a new enough
+// term and index.
+type NodeStatus string
+
+const (
+	NodeUnknown NodeStatus = "unknown"
+	NodeAlive   NodeStatus = "alive"
+	NodeFailed  NodeStatus = "failed"
+	NodeLeft    NodeStatus = "left"
+)
+
+type NodeType string
+
+const (
+	NodeVoter NodeType = "voter"
+)
+
+// Config represents all the tunables of autopilot
+type Config struct {
+	// CleanupDeadServers controls whether to remove dead servers when a new
+	// server is added to the Raft peers.
+	CleanupDeadServers bool
+
+	// LastContactThreshold is the limit on the amount of time a server can go
+	// without leader contact before being considered unhealthy.
+	LastContactThreshold time.Duration
+
+	// MaxTrailingLogs is the amount of entries in the Raft Log that a server can
+	// be behind before being considered unhealthy.
+	MaxTrailingLogs uint64
+
+	// MinQuorum sets the minimum number of servers required in a cluster
+	// before autopilot can prune dead servers.
+	MinQuorum uint
+
+	// ServerStabilizationTime is the minimum amount of time a server must be
+	// in a stable, healthy state before it can be added to the cluster. Only
+	// applicable with Raft protocol version 3 or higher.
+	ServerStabilizationTime time.Duration
+
+	Ext interface{}
```

Comment:
Right it is unused. It just fit with the pattern of riding promoter specific information in the Ext field of these structs. I guess I could add a method to the promoter to retrieve the promoter specific configuration. I will try that out and see how it feels.

Reply:
So another comment you left made me think about this more and we really need the promoter specific configuration to come through in the main autopilot configuration when autopilot retrieves it from the delegate.

For each round of reconciliation/state updating we must only get the whole configuration once. Otherwise the configuration can change on us leaving us with an inconsistent State and seemingly random promotions/demotions.

Reply:
I haven't had a chance to look at the updates, but at least in the version I looked at there were multiple calls to `GetConfig`, so that property did not exist at the time.